packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts

```diff
@@ -0,0 +1,149 @@
+import {ArrayValue, NumberValue, Schema} from '@coveo/bueno';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {recentResults} from '../../app/reducers';
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {RecentResultsSection} from '../../state/state-sections';
+import {RecentResultsState} from '../../features/recent-results/recent-results-state';
+import {
+  clearRecentResults,
+  registerRecentResults,
+} from '../../features/recent-results/recent-results-actions';
+import {
+  validateInitialState,
+  validateOptions,
+} from '../../utils/validate-payload';
+import {logClearRecentResults} from '../../features/recent-results/recent-results-analytics-actions';
+import {Result} from '../../api/search/search/result';
+
+export interface RecentResultsListProps {
+  /**
+   * The initial state that should be applied to the `RecentResultsList` controller.
+   */
+  initialState?: RecentResultsListInitialState;
+  /**
+   * The configuration options that should be applied to the `RecentResultsList` controller.
+   */
+  options?: RecentResultsListOptions;
+}
+
+const defaultRecentResultsProps: Required<RecentResultsListProps> = {
+  initialState: {
+    results: [],
+  },
+  options: {
+    maxLength: 10,
+  },
+};
+
+export interface RecentResultsListInitialState {
+  /**
+   * The list of recent results.
+   * @defaultValue `[]`
+   */
+  results: Result[];
+}
+
+const initialStateSchema = new Schema<RecentResultsListInitialState>({
+  results: new ArrayValue({required: true}),
+});
+
+export interface RecentResultsListOptions {
+  /**
+   * The maximum number of results to retain in the list.
+   * @defaultValue `10`
+   */
+  maxLength: number;
+}
+
+const optionsSchema = new Schema<RecentResultsListOptions>({
+  maxLength: new NumberValue({required: true, min: 1}),
+});
+
+/**
+ * The `RecentResultsList` controller manages the user's recent results.
+ */
+export interface RecentResultsList extends Controller {
+  /**
+   * The state of the RecentResultsList controller.
+   * */
+  state: RecentResultsState;
+  /**
+   * Clears the recent results list.
+   */
+  clear(): void;
+}
+
+export function validateRecentResultsProps(
+  engine: SearchEngine,
+  props?: RecentResultsListProps
+) {
+  validateOptions(
+    engine,
+    optionsSchema,
+    props?.options,
+    'buildRecentResultsList'
+  );
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props?.initialState,
+    'buildRecentResultsList'
+  );
+}
+
+/**
+ * Creates a `RecentResultsList` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configuration `RecentResultsList` properties.
+ * @returns A `RecentResultsList` controller instance.
+ * */
+export function buildRecentResultsList(
+  engine: SearchEngine,
+  props?: RecentResultsListProps
+): RecentResultsList {
+  if (!loadRecentResultsListReducer(engine)) {
```

Comment:
This check is a tautology, `loadRecentResultsListReducer` can only produce true values

Reply:
Every single controller does this exact thing. @samisayegh thoughts?

Reply:
The function always returns `true`. We use it as a type-guard to tell typescript that the engine in the `else` branch has a `RecentResultsSection`. Maybe there is a better way to do this? I prefer to avoid casts though,

Reply:
Ah thanks it makes more sense now. I don't know if there's a better way but I'd rather not modify every other controller in this branch if we found one now 😅